gillespy2/core/timespan.py

```diff
@@ -0,0 +1,119 @@
+GNU General Public License for more details.
+
+You should have received a copy of the GNU General Public License
+along with this program.  If not, see <http://www.gnu.org/licenses/>.
+"""
+import numpy as np
+from collections.abc import Iterator
+
+from gillespy2.core.jsonify import Jsonify
+from .gillespyError import TimespanError
+
+class TimeSpan(Iterator, Jsonify):
+    """
+    Model timespan that describes the duration to rub the simulation and at which timepoint to sample
+    the species populations during the simulation.
+
+    :param items: Evenly-spaced list of times at which to sample the species populations during the simulation. 
+            Best to use the form np.linspace(<start time>, <end time>, <number of time-points, inclusive>)
+    :type items: list, tuple, range, or numpy.ndarray
+    """
+    def __init__(self, items):
+        if isinstance(items, np.ndarray):
+            self.items = items
+        elif isinstance(items, (list, tuple, range)):
+            self.items = np.array(items)
+        else:
+            raise TimespanError("Timespan must be of type: list, tuple, range, or numpy.ndarray.")
+
+        self.validate()
+
+    def __eq__(self, o):
+        return self.items.__eq__(o).all()
+
+    def __getitem__(self, key):
+        return self.items.__getitem__(key)
+
+    def __iter__(self):
+        return self.items.__iter__()
+
+    def __len__(self):
+        return self.items.__len__()
+
+    def __next__(self):
+        return self.items.__next__()
+
+    @classmethod
+    def linspace(cls, t=None, num_points=None):
+        """
+        Creates a timespan using the form np.linspace(0, <t>, <num_points, inclusive>).
+
+        :param t: End time for the simulation.
+        :type t: int
+
+        :param num_points: Number of sample points for the species populations during the simulation.
+        :type num_points: int
+        """
+        if t is not None and not isinstance(t, int):
+            raise TimespanError("t must be of type int.")
+        if num_points is not None and not isinstance(num_points, int):
+            raise TimespanError("num_points must be of type int.")
+
+        if t is None:
+            t = 20
+        if num_points is None:
+            num_points = int(t / 0.05) + 1
+        items = np.linspace(0, t, num_points)
+        return cls(items)
+
+    @classmethod
+    def arange(cls, increment, t=None):
+        """
+        Creates a timespan using the form np.arange(0, <t, inclusive>, <increment>).
+
+        :param increment: Distance between sample points for the species populations during the simulation.
+        :type increment: float | int
+
+        :param t: End time for the simulation.
+        :type t: int
+        """
+        if t is not None and not isinstance(t, int):
+            raise TimespanError("t must be of type int.")
+        if not isinstance(increment, (float, int)):
+            raise TimespanError("increment must be of type float or int.")
```

Comment:
Ditto, this should have validation to assert that both `t` and `increment` are positive as any `numpy.arange` call with negative arguments results in an empty array.